util/no_os_pid.c

```diff
@@ -0,0 +1,145 @@
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <errno.h>
+#include "no_os_pid.h"
+#include "no_os_print_log.h"
+
+struct no_os_pid {
+	int iacc; // integral accumulator
+	int dacc; // derivative accumulator
+	int64_t output; // on 64-bits to avoid overflow
+	struct no_os_pid_config config; // copy of the user-provided configuration
+};
+
+/* Initialize a PID controller with given configuration. */
+int no_os_pid_init(struct no_os_pid **pid, struct no_os_pid_config config)
+{
+	if (!pid)
+		return -EINVAL;
+
+	if (config.output_clip.high < config.output_clip.low)
+		return -EINVAL;
+
+	*pid = calloc(1, sizeof(struct no_os_pid));
+	if (!*pid)
+		return -ENOMEM;
+
+	(*pid)->config = config;
+	(*pid)->output = config.initial;
+	(*pid)->iacc = 0;
+	(*pid)->dacc = 0;
+
+	return 0;
+}
+
+/* Perform PID control given a controller, a set-point and a process variable.
+ * This has to be called several times, until output stabilizes.
+ * The PID output is returned by the function. */
+int no_os_pid_control(struct no_os_pid *pid, int SP, int PV)
+{
+	int ERR; // error
+	int64_t P; // proportional component
+	int64_t I; // integral component
+	int64_t D; // derivative component
+
+	if (!pid)
+		return 0;
```

Comment:
do we want to return a success code here?

Reply:
So basically what this does, if user passes pid = 0 (the only wrong value we can detect, there can be countless others), the control function returns a constant output.

This is acceptable because it's the kind of error that you catch in development.

Although actually it should output not 0 but min(output_clip.low, output_clip.high).

Reply:
I guess I will change the function to return a status and the output will be a pointer argument...